File: apps/03-food-simple/backend/nodejs/src/services/llm/OpenAIService.js
import axios from 'axios';
import LLMInterface from './LLMInterface.js';

class OpenAIService extends LLMInterface {
  constructor() {
    super();
    this.apiKey = process.env.OPENAI_API_KEY;
    this.model = process.env.OPENAI_MODEL || 'gpt-3.5-turbo';
    this.axios = null;
    this.initializeAxios();
  }

  async initializeAxios() {
    this.axios = axios.create({
      baseURL: 'https://api.openai.com/v1',
      headers: {
        'Authorization': `Bearer ${this.apiKey}`,
        'Content-Type': 'application/json',
      },
      timeout: 30000,
    });
  }

  async sendMessage(message) {
    if (!this.apiKey) {
      throw new Error('OpenAI API key is not configured');
    }

    try {
      const response = await this.axios.post('/chat/completions', {
        model: this.model,
        messages: [
          {
            role: 'user',
            content: message
          }
        ],
        temperature: 0.7,
        max_tokens: 500,
      });

      if (!response.data.choices || response.data.choices.length === 0) {
        throw new Error('No response choices returned from OpenAI API');
      }

      return response.data.choices[0].message.content;
    } catch (error) {
      if (error.response?.status === 401) {
        throw new Error('OpenAI API key is invalid. Please check your API key and try again.');
      }
      throw new Error(`Error communicating with OpenAI API: ${error.message}`);
    }
  }

  getProvider() {
    return 'OpenAI';
  }

  getModel() {
    return this.model;
  }
}

export default OpenAIService; 

File: apps/03-food-simple/backend/python/src/services/llm/openai_service.py
import os
import httpx
from .llm_interface import LLMInterface

class OpenAIService(LLMInterface):
    """OpenAI service implementation"""
    
    def __init__(self):
        self.api_key = os.getenv("OPENAI_API_KEY")
        self.model = os.getenv("OPENAI_MODEL", "gpt-3.5-turbo")
        self.embedding_model = os.getenv("OPENAI_EMBEDDING_MODEL", "text-embedding-3-small")
        self.base_url = "https://api.openai.com/v1"
        self.client = httpx.AsyncClient(
            base_url=self.base_url,
            headers={
                "Authorization": f"Bearer {self.api_key}",
                "Content-Type": "application/json"
            },
            timeout=30.0
        )
        
        if not self.api_key:
            raise Exception("OpenAI API key is not configured")
    
    async def send_message(self, message: str) -> str:
        """Send a message to the OpenAI API and get a response"""
        try:
            response = await self.client.post(
                "/chat/completions",
                json={
                    "model": self.model,
                    "messages": [
                        {
                            "role": "user",
                            "content": message
                        }
                    ],
                    "temperature": 0.7,
                    "max_tokens": 1000
                }
            )
            
            if response.status_code != 200:
                raise Exception(f"OpenAI API error: {response.text}")
            
            data = response.json()
            
            if not data.get("choices") or len(data["choices"]) == 0:
                raise Exception(f"No response choices returned from OpenAI API: {data}")
            
            return data["choices"][0]["message"]["content"]
        except Exception as e:
            raise Exception(f"Error communicating with OpenAI API: {str(e)}")
    
    async def get_embeddings(self, text: str) -> list:
        """Get embeddings for the given text using the embedding model"""
        try:
            response = await self.client.post(
                "/embeddings",
                json={
                    "model": self.embedding_model,
                    "input": text
                }
            )
            
            if response.status_code != 200:
                raise Exception(f"OpenAI API error: {response.text}")
            
            data = response.json()
            
            if not data.get("data") or len(data["data"]) == 0:
                raise Exception(f"No embedding data returned from OpenAI API: {data}")
            
            return data["data"][0]["embedding"]
        except Exception as e:
            raise Exception(f"Error getting embeddings from OpenAI API: {str(e)}")
    
    def get_provider(self) -> str:
        """Get the name of the LLM provider"""
        return "OpenAI"
    
    def get_model(self) -> str:
        """Get the model name used by the LLM provider"""
        return self.model
        
    def get_embedding_model(self) -> str:
        """Get the embedding model name used by the LLM provider"""
        return self.embedding_model 